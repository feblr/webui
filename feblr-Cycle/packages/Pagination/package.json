{
  "name": "feblr-cycle-pagination",
  "version": "0.0.1",
  "description": "pagination for cycle.js",
  "main": "index.js",
  "scripts": {
    "test": "npm test",
    "build": "tsc -w ./src/index.ts"
  },
  "repository": {
    "type": "git",
    "url": "git+https://github.com/JacobChang/feblr-web.git"
  },
  "keywords": [
    "cycle.js",
    "pagination"
  ],
  "author": "Jacob Chang",
  "license": "ISC",
  "bugs": {
    "url": "https://github.com/JacobChang/feblr-web/issues"
  },
  "homepage": "https://github.com/JacobChang/feblr-web#readme",
  "devDependencies": {
    "typescript": "^2.1.4"
  },
  "dependencies": {
    "@cycle/dom": "^14.2.0",
    "xstream": "^9.0.0"
  }
}
