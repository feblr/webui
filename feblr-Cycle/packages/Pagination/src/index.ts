import xs, { Stream } from 'xstream';
import { DOMSource, VNode } from '@cycle/dom';
import { Pagination } from 'feblr-view-model';
import intent from './intent';
import update from './update';
import view from './view';

export interface ISources {
  dom: DOMSource;
  props$: Stream<Pagination>;
}

export default function PaginationComp(sources: ISources): Stream<VNode> {
  let actions$ = intent(sources);
  let models$ = update({
    actions$: actions$,
    props$: sources.props$
  });

  return view(models$);
}
