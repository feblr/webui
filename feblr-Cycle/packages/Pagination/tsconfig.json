{
  "compilerOptions": {
    "removeComments": false,
    "preserveConstEnums": true,
    "sourceMap": true,
    "declaration": true,
    "noImplicitAny": true,
    "suppressImplicitAnyIndexErrors": true,
    "moduleResolution": "node",
    "target": "ES6",
    "watch": true
  },
  "formatCodeOptions": {
    "indentSize": 2,
    "tabSize": 2
  },
   "compileOnSave": true
}
