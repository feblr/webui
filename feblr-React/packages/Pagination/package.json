{
  "name": "feblr-react-pagination",
  "version": "0.0.1",
  "description": "",
  "main": "dist/index.js",
  "scripts": {
    "test": "jest",
    "build": "babel ./src --out-dir ./dist --source-maps --presets es2015,react"
  },
  "author": "",
  "license": "ISC",
  "devDependencies": {
    "babel": "^6.5.2",
    "babel-cli": "^6.18.0",
    "babel-core": "^6.18.2",
    "babel-jest": "^17.0.2",
    "babel-loader": "^6.2.8",
    "babel-preset-es2015": "^6.18.0",
    "babel-preset-react": "^6.16.0",
    "babel-preset-stage-0": "^6.16.0",
    "enzyme": "^2.6.0",
    "jest": "^17.0.3",
    "react": "^15.4.1",
    "react-addons-test-utils": "^15.4.1",
    "react-dom": "^15.4.1",
    "react-test-renderer": "^15.4.1"
  },
  "jest": {
    "unmockedModulePathPatterns": [
      "node_modules/react/",
      "node_modules/enzyme/"
    ]
  }
}
