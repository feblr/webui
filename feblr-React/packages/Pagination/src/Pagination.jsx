import { Pagination as FeblrPagination } from 'feblr-view-model';
import React, { Component, PropTypes } from 'react';

export default class Pagination extends Component {
  static propTypes = {
    onPageChange: PropTypes.func,
    feblrPagination: PropTypes.instanceOf(FeblrPagination)
  };

  constructor(props) {
    super(props);

    this.gotoPrevPage = this.gotoPrevPage.bind(this);
    this.gotoNextPage = this.gotoNextPage.bind(this);
    this.gotoPage = this.gotoPage.bind(this);
    this.updateJumper = this.updateJumper.bind(this);

    this.state = {
      jumper: {
        page: 1
      }
    }
  }

  gotoPrevPage() {
    this.props.onPageChange({
      page: this.props.feblrPagination.currPage - 1
    });
  }

  gotoNextPage() {
    this.props.onPageChange({
      page: this.props.feblrPagination.currPage + 1
    });
  }

  gotoPage(event) {
    var targetPage = this.props.feblrPagination.currPage;
    if (event.target.tagName === 'BUTTON') {
      targetPage = parseInt(event.target.textContent) - 1;
    } else {
      targetPage = parseInt(this.state.jumper.page) - 1;
    }

    this.props.onPageChange({
      page: targetPage
    });
  }

  updateJumper(event) {
    this.setState({
      jumper: {
        page: parseInt(event.target.value)
      }
    });
  }

  render() {
    var pages = this.props.feblrPagination.pages.map(page => {
      return (
        <li className="page" key={page.index} >
          <button type="button" onClick={this.gotoPage}>{page.index + 1}</button>
        </li>
      )
    });

    return (
      <div className="feblr-pagination">
        <ul>
          <li className="prev">
            <button type="button" onClick={this.gotoPrevPage}><i className="feblr-icon-arrow-left icon"></i></button>
          </li>
          {pages}
          <li className="next">
            <button type="button" onClick={this.gotoNextPage}><i className="feblr-icon-arrow-right icon"></i></button>
          </li>
        </ul>
        <div className="jumper">
          <form onSubmit={this.gotoPage}>
            <input type="number" value={this.state.jumper.page} onChange={this.updateJumper} />
          </form>
        </div>
      </div>
    )
  }
}
