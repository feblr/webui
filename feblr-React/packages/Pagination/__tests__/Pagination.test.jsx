// Link.react-test.js
import React from 'react';
import { shallow } from 'enzyme';
import renderer from 'react-test-renderer';
import Pagination from '../src/index';
import { Pagination as FeblrPagination } from 'feblr-view-model';

test('Pagination structure', () => {
  let pagination = new FeblrPagination(0, 10, 100);
  let callback = jest.fn();

  const component = renderer.create(
    <Pagination feblrPagination={pagination} onPageChange={callback} />
  );
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

test('Pagination next page', () => {
  let pagination = new FeblrPagination(0, 10, 100);
  let callback = jest.fn();

  const component = shallow(
    <Pagination feblrPagination={pagination} onPageChange={callback} />
  );

  component.find('.prev button').simulate('click');

  expect(callback.mock.calls.length).toEqual(1);
  expect(callback.mock.calls).toEqual([[{ page: -1 }]]);
});

test('Pagination next page', () => {
  let pagination = new FeblrPagination(0, 10, 100);
  let callback = jest.fn();

  const component = shallow(
    <Pagination feblrPagination={pagination} onPageChange={callback} />
  );

  component.find('.next button').simulate('click');

  expect(callback.mock.calls.length).toEqual(1);
  expect(callback.mock.calls).toEqual([[{ page: 1 }]]);
});

test('Pagination target page', () => {
  let pagination = new FeblrPagination(0, 10, 100);
  let callback = jest.fn();

  const component = shallow(
    <Pagination feblrPagination={pagination} onPageChange={callback} />
  );

  let index = 2;
  component.find('.page button').at(index).simulate('click', {
    target: {
      tagName: 'BUTTON',
      textContent: index + 1 + ''
    }
  });

  expect(callback.mock.calls.length).toEqual(1);
  expect(callback.mock.calls).toEqual([[{ page: index }]]);
});

test('Pagination jumper', () => {
  let pagination = new FeblrPagination(0, 10, 100);
  let callback = jest.fn();

  const component = shallow(
    <Pagination feblrPagination={pagination} onPageChange={callback} />
  );

  let input = '3';
  component.find('.jumper input').first().simulate('change', {
    target: {
      value: input
    }
  });

  component.find('.jumper form').simulate('submit', {
    target: component.find('.jumper form')
  });

  expect(callback.mock.calls.length).toEqual(1);
  expect(callback.mock.calls).toEqual([[{ page: parseInt(input) - 1 }]]);
});
