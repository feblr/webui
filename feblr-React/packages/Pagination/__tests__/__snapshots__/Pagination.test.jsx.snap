exports[`test Pagination structure 1`] = `
<div
  className="feblr-pagination">
  <ul>
    <li
      className="prev">
      <button
        onClick={[Function]}
        type="button">
        <i
          className="feblr-icon-arrow-left icon" />
      </button>
    </li>
    <li
      className="page">
      <button
        onClick={[Function]}
        type="button">
        1
      </button>
    </li>
    <li
      className="page">
      <button
        onClick={[Function]}
        type="button">
        2
      </button>
    </li>
    <li
      className="page">
      <button
        onClick={[Function]}
        type="button">
        3
      </button>
    </li>
    <li
      className="page">
      <button
        onClick={[Function]}
        type="button">
        4
      </button>
    </li>
    <li
      className="page">
      <button
        onClick={[Function]}
        type="button">
        5
      </button>
    </li>
    <li
      className="next">
      <button
        onClick={[Function]}
        type="button">
        <i
          className="feblr-icon-arrow-right icon" />
      </button>
    </li>
  </ul>
  <div
    className="jumper">
    <form
      onSubmit={[Function]}>
      <input
        onChange={[Function]}
        type="number"
        value={1} />
    </form>
  </div>
</div>
`;
